### PR TITLE
First steps into microservices world.

### PR DESCRIPTION
**Pull request by RYelzhan**
Implemented all the necessary requirements. However, not all recommendations are still implemented like Zipkin and Sleuth for tracing and transfer of authentication layer into ApiGateway. Will add these in the next few days.

Key Features and Improvements:

1. Refactoring and Security Enhancements:
- Moved the main application to its module.
- Transitioned authenticated user retrieval to the Spring Security context for improved security and maintainability.
- Customized security-related response messages for better clarity and user experience.

2. Eureka Server Setup:
- Added Eureka Server for service registration and discovery.

3. Microservice Communication:
- Enabled and tested communication between SpringGymApplication and Statistics.
- Verified training addition and summary retrieval across applications.

4. Synchronization and Modularization:
- Synchronized training deletion between GymSpringCore and Statistics using Feign clients.
- Consolidated shared classes into a new GymCommon module to enhance code reusability.
- Decoupled the Statistics service from direct dependencies on GymSpringCore.

5. Centralized Communication:
- Centralized interactions between GymSpringCore and Statistics services into a single service class.
- Tested the functionality to ensure seamless communication.

6. Resilience with Circuit Breaker:
- Implemented the Circuit Breaker pattern.
- Gracefully handle unavailability of the Statistics Service, ensuring application robustness.

7. API Gateway Integration:
- Created a new module for the API Gateway to streamline service access.
- Enabled basic request distribution through the API Gateway.
- Configured service routing with URL prefixes /gym and /stats for respective services.

8. Transaction ID Handling:
- Refactored the creation of transaction IDs.
- Ensured proper transfer of transaction IDs to downstream services for consistent request tracing.

Testing and Validation

1. Controller Test Coverage:
- Covered edge cases for AuthenticationController, TrainerController, TraineeController, and UserController with comprehensive tests.

2. New Functionalities Validation:
- Verified functionality for API Gateway, Circuit Breaker, and synchronization features.
- Conducted integration tests for communication and data consistency between GymSpringCore and Statistics.

------------------------------------------------------

This pull request delivers a significant enhancement in service resilience, modularization, and security, along with robust test coverage for critical application controllers.

### CODE REVIEW

File: GymSpringCore/src/main/java/com/epam/wca/gym/communication/impl/StatisticsCommunicationServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+        throw new ServiceUnavailableException(STATISTICS_SERVICE_UNAVAILABLE_USER_MESSAGE);
+    }
+
+    public void fallbackResponse(TrainerTrainingAddDTO trainingAddDTO, Throwable e) {
+        log.error("TransactionId: {}, Fallback for getWorkload: {}",
+                TransactionContext.getTransactionId(), e.getMessage());
+
+        throw new ServiceUnavailableException(STATISTICS_SERVICE_UNAVAILABLE_USER_MESSAGE);
+    }
```
Comment (maximelyashevich) [inline GymSpringCore/src/main/java/com/epam/wca/gym/communication/impl/StatisticsCommunicationServiceImpl.java:58]: Please reduce code duplication.

File: GymSpringCore/src/test/java/com/epam/wca/gym/controller/TrainerControllerImplTest.java [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+                             "lastName": "Doe",
+                             "trainingType": "YOGA",
+                             "isActive": true
+                        }
+                    """))
+                .andExpect(status().isBadRequest());
+    }
+
+    @Test
+    void testUpdateProfile_FirstnameTooShort_ShouldReturnBadRequest() throws Exception {
```
Comment (maximelyashevich) [inline GymSpringCore/src/test/java/com/epam/wca/gym/controller/TrainerControllerImplTest.java:94]: Similar tests should be grouped in a single `Parameterized` test.

See an example:
```Java
@ParameterizedTest
    @CsvSource({
        "1, 100",
        "2, 200",
        "3, 300",
    })
    void testLevels(int level, int health) {
        setLevel(level);
        runGame();
        assertEquals(playerHealth(), health);
    }
```

File: GymSpringCore/src/test/java/com/epam/wca/gym/transaction/TransactionManagementTest.java [hunk 3]
```diff
@@ -25,7 +18,7 @@ void contextLoads() {
     }
 
     // TODO: finish it
-    @ParameterizedTest
+/*    @ParameterizedTest
```
Comment (maximelyashevich) [inline GymSpringCore/src/test/java/com/epam/wca/gym/transaction/TransactionManagementTest.java:21]: Remove commented code.

**Review by maximelyashevich — APPROVED** [review]
Reviewed & approved from my side. Good job!

File: GymSpringCore/src/main/java/com/epam/wca/gym/communication/impl/StatisticsCommunicationServiceImpl.java (hunk 1 above)
Comment (RYelzhan) [reply GymSpringCore/src/main/java/com/epam/wca/gym/communication/impl/StatisticsCommunicationServiceImpl.java:58]: Done.

File: GymSpringCore/src/test/java/com/epam/wca/gym/controller/TrainerControllerImplTest.java (hunk 2 above)
Comment (RYelzhan) [reply GymSpringCore/src/test/java/com/epam/wca/gym/controller/TrainerControllerImplTest.java:94]: Done. For every test of controllers.

File: GymSpringCore/src/main/java/com/epam/wca/gym/communication/impl/StatisticsCommunicationServiceImpl.java (hunk 1 above)
Comment (RYelzhan) [reply GymSpringCore/src/main/java/com/epam/wca/gym/communication/impl/StatisticsCommunicationServiceImpl.java:58]: Done for all controller methods.

File: GymSpringCore/src/test/java/com/epam/wca/gym/controller/TrainerControllerImplTest.java (hunk 2 above)
Comment (RYelzhan) [reply GymSpringCore/src/test/java/com/epam/wca/gym/controller/TrainerControllerImplTest.java:94]: Done for all controller methods.

File: GymSpringCore/src/test/java/com/epam/wca/gym/transaction/TransactionManagementTest.java (hunk 3 above)
Comment (RYelzhan) [reply GymSpringCore/src/test/java/com/epam/wca/gym/transaction/TransactionManagementTest.java:21]: Done.

**Review by Atabek04 — APPROVED** [review]
Noting to add from my side, lgtm! Great job!
